### ISSUE TITLE
Why not average over batch dimension?

### ISSUE THREAD

**Issue by CXX1113** [altosaar/variational-autoencoder#16]
in train_variational_autoencoder_pytorch.py file 

# sum over batch dimension
loss = -elbo.sum(0)

change into
# average over batch dimension
loss = -elbo.mean(0)

**Comment by jaanaltosaar** [comment]
My goal was to have the math match the code as closely as possible. The ELBO is defined as a sum, not an average.

It won't make a difference with adaptive learning rate optimizers like Adam and RMSProp (the constant cancels out in the normalization).

**Comment by CXX1113** [comment]
In fact, ELBO is defined for a single sample, I can't find a definition for ELBO for all samples in a batch. So you said "The ELBO is defined as a sum, not an average." Where is the source?

**Comment by jaanaltosaar** [comment]
@alanlisten - Eq. 13 is the ELBO here: https://arxiv.org/pdf/1601.00670.pdf

The first term is the expectation of the log joint of the model.

The log joint of the model decomposes into the log likelihood and the log prior. 

The log-likelihood of all the datapoints decomposes into the sum of the likelihood of every datapoint. There is no average. As an example, plug in Eq. 7: the product decomposes into a sum when you apply the logarithm to the log joint of the model.

Let me know if you still have questions! I still find some of this stuff confusing :)

**Comment by CXX1113** [comment]
Thank you for your patience.
https://arxiv.org/pdf/1312.6114.pdf. 
This is an authoritative paper on VAE. 
In this paper, according to Eq.3, ELBO is for a single data point (i.e. x^(i)). You said that when you want to log the joint probability, it is about all the data points.I don't understand very well.

**Comment by jaanaltosaar** [comment]
Above Eq. 1 the text says the marginal likelihood is a sum of the likelihoods of all datapoints. 

This means that Eq. 3 is provides a lower bound on the marginal likelihood of a single datapoint. To get a lower bound on all datapoints, you have to sum the inequalities. (There is no average.)

**Comment by CXX1113** [comment]
I ignored the above sentence of Eq.1, I think you are right. Thank you for letting me know more about VAE, thank you very much again. O(∩_∩)O
